stl/inc/ranges

```diff
@@ -620,6 +620,132 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
+    // CLASS TEMPLATE ranges::_Non_propagating_cache
+    template <class _Ty>
+    concept _Objectable = is_object_v<_Ty>;
+
+    // clang-format off
+    template <_Objectable _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20_DYNALLOC ~_Non_propagating_cache() {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy = {};
+            }
+        }
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Dummy   = {};

```

Comment:
This should be ill-formed at compile-time / UB at runtime since `_Dummy` isn't the active alternative of `_Other`'s anonymous union member.
```suggestion
                _Construct_in_place(_Other._Dummy);
```

Reply:
Wasn't that allowed by WG21-P1330? If it works with `float` (as shown in the example in the paper) shouldn't it work with all trivially copyable types? Is it really UB in that case?

Reply:
WG21-P1330 allows assignments that change the active alternative in constant expressions. The issue is that this assignment doesn't change the active alternative since `_Dummy` has a non-trivial default constructor (https://eel.is/c++draft/class.union.general#6).